### PR TITLE
feat: Better SLA missed events popover 

### PR DESCRIPTION
**Pull request by muhsin-k**
Fixes https://linear.app/chatwoot/issue/CW-3198/update-the-popover-ui-for-long-list-of-misses

**Light theme**

CleanShot 2024-04-09 at 16 29 59@2x

**Dark theme**

CleanShot 2024-04-09 at 16 31 00@2x

### CODE REVIEW

**Comment by muhsin-k** [comment]
Preparing review...

**Comment by muhsin-k** [comment]
## PR Review

<table>
<tr>
<tr><td> ⏱️&nbsp;<strong>Estimated&nbsp;effort&nbsp;to&nbsp;review [1-5]</strong></td><td>

3, because the PR involves changes across multiple components including Vue components, constants, and localization files. The logic related to displaying SLA missed events has been refactored, requiring a careful review of both the logic and UI implications.


</td></tr>
<tr><td> 🧪&nbsp;<strong>Relevant tests</strong></td><td>

No


</td></tr>
<tr><td rowspan=2> 🔍&nbsp;<strong>Possible issues</strong></td>
<td>

<strong>Possible Bug:</strong> The `v-if` conditions in `SLAPopoverCard.vue` for `frtMisses`, `nrtMisses`, and `rtMisses` might always evaluate to true because computed properties in Vue are truthy even when the underlying array is empty. This could lead to rendering empty `SLA-event-item` components.</td></tr>
<tr>
<td>

<strong>Performance Concern:</strong> In `SLAPopoverCard.vue`, filtering `slaMissedEvents` for each type (FRT, NRT, RT) on every re-render might be inefficient, especially if the list is large. Consider memoizing these operations or computing them less frequently.</td></tr>
<tr><td> 🔒&nbsp;<strong>Security concerns</strong></td><td>

No

</td></tr>
</table>


<details><summary> <strong>Code feedback:</strong></summary>

<hr><table><tr><td>relevant file</td><td>app/javascript/dashboard/components/widgets/conversation/components/SLAPopoverCard.vue
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Consider using a computed property that checks if the arrays `frtMisses`, `nrtMisses`, and `rtMisses` are empty before rendering the `SLA-event-item` components. This will prevent rendering empty components. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/chatwoot/chatwoot/pull/9215/files#diff-18884c51f04ded895149c33038da6599e8bcc9bd2cf86df822a62d18c007b2baR48'>v-if="frtMisses"</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>app/javascript/dashboard/components/widgets/conversation/components/SLAPopoverCard.vue
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Memoize the filtered SLA missed events or compute them once and react to specific changes to improve performance. This can be achieved by using a watcher or a more specific reactive property that triggers the computation. [medium]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/chatwoot/chatwoot/pull/9215/files#diff-18884c51f04ded895149c33038da6599e8bcc9bd2cf86df822a62d18c007b2baR18'>const frtMisses = computed(() =></a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>app/javascript/dashboard/components/widgets/conversation/components/SLAPopoverCard.vue
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

For better accessibility and user experience, consider adding an aria-label or similar attribute to the "Show More" and "Hide" buttons to provide more context to screen readers. [medium]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/chatwoot/chatwoot/pull/9215/files#diff-18884c51f04ded895149c33038da6599e8bcc9bd2cf86df822a62d18c007b2baR68'>@click="toggleShowAllNRT"</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>app/javascript/dashboard/components/widgets/conversation/components/SLAEventItem.vue
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Ensure the `formatDate` function handles invalid or undefined `timestamp` values gracefully to avoid runtime errors in the template rendering. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/chatwoot/chatwoot/pull/9215/files#diff-8c7ee27e9ecc210a096b6eb6c5d29ad61253b39a3061864340cda75f0bf04942R30'>{{ formatDate(item.created_at) }}</a></td></tr></table><hr>

</details><hr>

<details> <summary><strong>✨ Review tool usage guide:</strong></summary><hr> 

**Overview:**
The `review` tool scans the PR code changes, and generates a PR review which includes several types of feedbacks, such as possible PR issues, security threats and relevant test in the PR. More feedbacks can be [added](https://pr-agent-docs.codium.ai/tools/review/#general-configurations) by configuring the tool.

The tool can be triggered [automatically](https://pr-agent-docs.codium.ai/usage-guide/automations_and_usage/#github-app-automatic-tools-when-a-new-pr-is-opened) every time a new PR is opened, or can be invoked manually by commenting on any PR.
- When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L23) related to the review tool (`pr_reviewer` section), use the following template:
```
/review --pr_reviewer.some_config1=... --pr_reviewer.some_config2=...
```
- With a [configuration file](https://pr-agent-docs.codium.ai/usage-guide/configuration_options/), use the following template:
```
[pr_reviewer]
some_config1=...
some_config2=...
```
    

See the review [usage page](https://pr-agent-docs.codium.ai/tools/review/) for a comprehensive guide on using this tool.


</details>

**Comment by muhsin-k** [comment]
Preparing review...

File: app/javascript/dashboard/components/widgets/conversation/components/SLAEventItem.vue [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    <span
+      class="text-sm sticky top-0 h-fit font-normal tracking-[-0.6%] min-w-[140px] truncate text-slate-600 dark:text-slate-200"
+    >
+      {{ label }}
+    </span>
+    <div class="flex flex-col w-full gap-2">
+      <span
+        v-for="item in items"
+        :key="item.id"
+        class="text-sm font-normal text-slate-900 dark:text-slate-25 text-right"
```
Comment (scmmishra) [inline app/javascript/dashboard/components/widgets/conversation/components/SLAEventItem.vue:28]:
```suggestion
        class="text-sm font-normal text-slate-900 dark:text-slate-25 text-right tabular-nums"
```

We can use `tabular-nums` it'll fix the width of the numbers, make it easier to scan through by the users

**Review by scmmishra — APPROVED** [review]

**Pull request merged**